File: src/product-entry-program-companions.ts
type JsonRecord = Record<string, unknown>;

export interface ProgramCheckInput {
  check_id: string;
  title: string;
  status: string;
  blocking: boolean;
  summary: string;
  command: string;
}

export interface WorkflowCoverageItemInput {
  step_id: string;
  manual_flow_label: string;
  coverage_status: string;
  current_surface: string;
  remaining_gap: string;
}

export interface GuardrailClassInput {
  guardrail_id: string;
  trigger: string;
  symptom: string;
  recommended_command: string;
}

export interface ProgramStepInput {
  step_id: string;
  command: string;
  surface_kind: string;
  title?: string;
}

export interface ProgramSurfaceInput {
  surface_kind: string;
  command?: string;
  ref?: string;
}

export interface ClearanceTargetInput {
  target_id: string;
  title: string;
  commands: string[];
}

export interface ProgramCapabilityInput {
  capability_id: string;
  owner: string;
  summary: string;
}

export interface ProgramSequenceStepInput {
  step_id: string;
  phase_id: string;
  status: string;
  summary: string;
  title?: string;
}

export interface DetailedReadinessInput {
  surface_kind: string;
  verdict: string;
  usable_now: boolean;
  good_to_use_now: boolean;
  fully_automatic: boolean;
  user_experience_level: string;
  summary: string;
  recommended_start_surface: string;
  recommended_start_command: string;
  recommended_loop_surface: string;
  recommended_loop_command: string;
  workflow_coverage: WorkflowCoverageItemInput[];
  blocking_gaps: string[];
}

export interface ProductEntryPreflightInput {
  summary: string;
  recommended_check_command: string;
  recommended_start_command: string;
  checks: ProgramCheckInput[];
}

export interface ProductEntryGuardrailsInput {
  summary: string;
  guardrail_classes: GuardrailClassInput[];
  recovery_loop: ProgramStepInput[];
}

export interface ClearanceLaneInput {
  surface_kind: string;
  summary: string;
  recommended_step_id: string;
  recommended_command: string;
  clearance_targets: ClearanceTargetInput[];
  clearance_loop: ProgramStepInput[];
  proof_surfaces: ProgramSurfaceInput[];
  recommended_phase_command: string;
}

export interface BackendDeconstructionLaneInput {
  summary: string;
  substrate_targets: ProgramCapabilityInput[];
  backend_retained_now: string[];
  current_backend_chain: string[];
  optional_executor_proofs: JsonRecord[];
  promotion_rules: string[];
  deconstruction_map_doc: string;
  recommended_phase_command: string;
  surface_kind?: string;
}

export interface PlatformTargetInput {
  summary: string;
  sequence_scope: string;
  current_step_id: string;
  current_readiness_summary: string;
  north_star_topology: JsonRecord;
  target_internal_modules: string[];
  landing_sequence: ProgramSequenceStepInput[];
  completed_step_ids: string[];
  remaining_step_ids: string[];
  promotion_gates: string[];
  recommended_phase_command: string;
  surface_kind?: string;
  land_now?: string[];
  not_yet?: string[];
}

function optionalString(value: unknown) {
  return typeof value === 'string' && value.trim() ? value.trim() : null;
}

function requireString(value: unknown, field: string) {
  const text = optionalString(value);
  if (!text) {
    throw new Error(`product entry program companion 缺少字符串字段: ${field}`);
  }
  return text;
}

function requireBoolean(value: unknown, field: string) {
  if (typeof value !== 'boolean') {
    throw new Error(`product entry program companion 缺少布尔字段: ${field}`);
  }
  return value;
}

function requireRecord(value: unknown, field: string) {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(`product entry program companion 缺少对象字段: ${field}`);
  }
  return value as JsonRecord;
}

function readStringList(value: unknown, field: string) {
  if (!Array.isArray(value)) {
    throw new Error(`product entry program companion 缺少数组字段: ${field}`);
  }
  return value.map((entry, index) => requireString(entry, `${field}[${index}]`));
}

function normalizeProgramCheck(value: ProgramCheckInput, field: string) {
  return {
    check_id: requireString(value.check_id, `${field}.check_id`),
    title: requireString(value.title, `${field}.title`),
    status: requireString(value.status, `${field}.status`),
    blocking: requireBoolean(value.blocking, `${field}.blocking`),
    summary: requireString(value.summary, `${field}.summary`),
    command: requireString(value.command, `${field}.command`),
  };
}

function normalizeWorkflowCoverageItem(value: WorkflowCoverageItemInput, field: string) {
  return {
    step_id: requireString(value.step_id, `${field}.step_id`),
    manual_flow_label: requireString(value.manual_flow_label, `${field}.manual_flow_label`),
    coverage_status: requireString(value.coverage_status, `${field}.coverage_status`),
    current_surface: requireString(value.current_surface, `${field}.current_surface`),
    remaining_gap: requireString(value.remaining_gap, `${field}.remaining_gap`),
  };
}

function normalizeGuardrailClass(value: GuardrailClassInput, field: string) {
  return {
    guardrail_id: requireString(value.guardrail_id, `${field}.guardrail_id`),
    trigger: requireString(value.trigger, `${field}.trigger`),
    symptom: requireString(value.symptom, `${field}.symptom`),
    recommended_command: requireString(value.recommended_command, `${field}.recommended_command`),
  };
}

function normalizeProgramStep(value: ProgramStepInput, field: string) {
  return {
    step_id: requireString(value.step_id, `${field}.step_id`),
    command: requireString(value.command, `${field}.command`),
    surface_kind: requireString(value.surface_kind, `${field}.surface_kind`),
    ...(optionalString(value.title) ? { title: optionalString(value.title)! } : {}),
  };
}

function normalizeProgramSurface(value: ProgramSurfaceInput, field: string) {
  const command = optionalString(value.command);
  const ref = optionalString(value.ref);
  if (!command && !ref) {
    throw new Error('product entry program surface 至少要提供 command 或 ref。');
  }
  return {
    surface_kind: requireString(value.surface_kind, `${field}.surface_kind`),
    ...(command ? { command } : {}),
    ...(ref ? { ref } : {}),
  };
}

function normalizeClearanceTarget(value: ClearanceTargetInput, field: string) {
  return {
    target_id: requireString(value.target_id, `${field}.target_id`),
    title: requireString(value.title, `${field}.title`),
    commands: readStringList(value.commands, `${field}.commands`),
  };
}

function normalizeProgramCapability(value: ProgramCapabilityInput, field: string) {
  return {
    capability_id: requireString(value.capability_id, `${field}.capability_id`),
    owner: requireString(value.owner, `${field}.owner`),
    summary: requireString(value.summary, `${field}.summary`),
  };
}

function normalizeProgramSequenceStep(value: ProgramSequenceStepInput, field: string) {
  return {
    step_id: requireString(value.step_id, `${field}.step_id`),
    phase_id: requireString(value.phase_id, `${field}.phase_id`),
    status: requireString(value.status, `${field}.status`),
    summary: requireString(value.summary, `${field}.summary`),
    ...(optionalString(value.title) ? { title: optionalString(value.title)! } : {}),
  };
}

export function buildProgramCheck(input: ProgramCheckInput) {
  return normalizeProgramCheck(input, 'program_check');
}

export function buildProductEntryPreflight(input: ProductEntryPreflightInput) {
  const checks = input.checks.map((check, index) => normalizeProgramCheck(check, `checks[${index}]`));
  const blockingCheckIds = checks
    .filter((check) => check.blocking && check.status !== 'pass')
    .map((check) => check.check_id);
  return {
    surface_kind: 'product_entry_preflight',
    summary: requireString(input.summary, 'summary'),
    ready_to_try_now: blockingCheckIds.length === 0,
    recommended_check_command: requireString(input.recommended_check_command, 'recommended_check_command'),
    recommended_start_command: requireString(input.recommended_start_command, 'recommended_start_command'),
    blocking_check_ids: blockingCheckIds,
    checks,
  };
}

export function buildWorkflowCoverageItem(input: WorkflowCoverageItemInput) {
  return normalizeWorkflowCoverageItem(input, 'workflow_coverage_item');
}

export function buildGuardrailClass(input: GuardrailClassInput) {
  return normalizeGuardrailClass(input, 'guardrail_class');
}

export function buildProgramStep(input: ProgramStepInput) {
  return normalizeProgramStep(input, 'program_step');
}

export function buildProgramSurface(input: ProgramSurfaceInput) {
  return normalizeProgramSurface(input, 'program_surface');
}

export function buildClearanceTarget(input: ClearanceTargetInput) {
  return normalizeClearanceTarget(input, 'clearance_target');
}

export function buildProgramCapability(input: ProgramCapabilityInput) {
  return normalizeProgramCapability(input, 'program_capability');
}

export function buildProgramSequenceStep(input: ProgramSequenceStepInput) {
  return normalizeProgramSequenceStep(input, 'program_sequence_step');
}

export function buildDetailedReadiness(input: DetailedReadinessInput) {
  return {
    surface_kind: requireString(input.surface_kind, 'surface_kind'),
    verdict: requireString(input.verdict, 'verdict'),
    usable_now: requireBoolean(input.usable_now, 'usable_now'),
    good_to_use_now: requireBoolean(input.good_to_use_now, 'good_to_use_now'),
    fully_automatic: requireBoolean(input.fully_automatic, 'fully_automatic'),
    user_experience_level: requireString(input.user_experience_level, 'user_experience_level'),
    summary: requireString(input.summary, 'summary'),
    recommended_start_surface: requireString(input.recommended_start_surface, 'recommended_start_surface'),
    recommended_start_command: requireString(input.recommended_start_command, 'recommended_start_command'),
    recommended_loop_surface: requireString(input.recommended_loop_surface, 'recommended_loop_surface'),
    recommended_loop_command: requireString(input.recommended_loop_command, 'recommended_loop_command'),
    workflow_coverage: input.workflow_coverage.map((item, index) => (
      normalizeWorkflowCoverageItem(item, `workflow_coverage[${index}]`)
    )),
    blocking_gaps: readStringList(input.blocking_gaps, 'blocking_gaps'),
  };
}

export function buildProductEntryGuardrails(input: ProductEntryGuardrailsInput) {
  return {
    surface_kind: 'product_entry_guardrails',
    summary: requireString(input.summary, 'summary'),
    guardrail_classes: input.guardrail_classes.map((entry, index) => (
      normalizeGuardrailClass(entry, `guardrail_classes[${index}]`)
    )),
    recovery_loop: input.recovery_loop.map((entry, index) => (
      normalizeProgramStep(entry, `recovery_loop[${index}]`)
    )),
  };
}

export function buildClearanceLane(input: ClearanceLaneInput) {
  return {
    surface_kind: requireString(input.surface_kind, 'surface_kind'),
    summary: requireString(input.summary, 'summary'),
    recommended_step_id: requireString(input.recommended_step_id, 'recommended_step_id'),
    recommended_command: requireString(input.recommended_command, 'recommended_command'),
    clearance_targets: input.clearance_targets.map((entry, index) => (
      normalizeClearanceTarget(entry, `clearance_targets[${index}]`)
    )),
    clearance_loop: input.clearance_loop.map((entry, index) => (
      normalizeProgramStep(entry, `clearance_loop[${index}]`)
    )),
    proof_surfaces: input.proof_surfaces.map((entry, index) => (
      normalizeProgramSurface(entry, `proof_surfaces[${index}]`)
    )),
    recommended_phase_command: requireString(input.recommended_phase_command, 'recommended_phase_command'),
  };
}

export function buildBackendDeconstructionLane(input: BackendDeconstructionLaneInput) {
  return {
    surface_kind: requireString(input.surface_kind ?? 'phase4_backend_deconstruction_lane', 'surface_kind'),
    summary: requireString(input.summary, 'summary'),
    substrate_targets: input.substrate_targets.map((entry, index) => (
      normalizeProgramCapability(entry, `substrate_targets[${index}]`)
    )),
    backend_retained_now: readStringList(input.backend_retained_now, 'backend_retained_now'),
    current_backend_chain: readStringList(input.current_backend_chain, 'current_backend_chain'),
    optional_executor_proofs: input.optional_executor_proofs.map((entry, index) => (
      { ...requireRecord(entry, `optional_executor_proofs[${index}]`) }
    )),
    promotion_rules: readStringList(input.promotion_rules, 'promotion_rules'),
    deconstruction_map_doc: requireString(input.deconstruction_map_doc, 'deconstruction_map_doc'),
    recommended_phase_command: requireString(input.recommended_phase_command, 'recommended_phase_command'),
  };
}

export function buildPlatformTarget(input: PlatformTargetInput) {
  const landingSequence = input.landing_sequence.map((entry, index) => (
    normalizeProgramSequenceStep(entry, `landing_sequence[${index}]`)
  ));
  const knownStepIds = new Set(landingSequence.map((entry) => entry.step_id));
  for (const [field, values] of [
    ['completed_step_ids', input.completed_step_ids],
    ['remaining_step_ids', input.remaining_step_ids],
  ] as const) {
    for (const stepId of readStringList(values, field)) {
      if (!knownStepIds.has(stepId)) {
        throw new Error(`platform target ${field} 必须引用 landing_sequence 中的 step_id。`);
      }
    }
  }

  return {
    surface_kind: requireString(input.surface_kind ?? 'phase5_platform_target', 'surface_kind'),
    summary: requireString(input.summary, 'summary'),
    sequence_scope: requireString(input.sequence_scope, 'sequence_scope'),
    current_step_id: requireString(input.current_step_id, 'current_step_id'),
    current_readiness_summary: requireString(input.current_readiness_summary, 'current_readiness_summary'),
    north_star_topology: { ...requireRecord(input.north_star_topology, 'north_star_topology') },
    target_internal_modules: readStringList(input.target_internal_modules, 'target_internal_modules'),
    landing_sequence: landingSequence,
    completed_step_ids: readStringList(input.completed_step_ids, 'completed_step_ids'),
    remaining_step_ids: readStringList(input.remaining_step_ids, 'remaining_step_ids'),
    promotion_gates: readStringList(input.promotion_gates, 'promotion_gates'),
    recommended_phase_command: requireString(input.recommended_phase_command, 'recommended_phase_command'),
    ...(input.land_now ? { land_now: readStringList(input.land_now, 'land_now') } : {}),
    ...(input.not_yet ? { not_yet: readStringList(input.not_yet, 'not_yet') } : {}),
  };
}

File: python/opl-harness-shared/src/opl_harness_shared/product_entry_program_companions.py
from __future__ import annotations

from typing import Any, Mapping, Sequence


def _non_empty_text(value: object) -> str | None:
    text = str(value or "").strip()
    return text or None


def _require_string(value: object, field: str) -> str:
    text = _non_empty_text(value)
    if text is None:
        raise ValueError(f"product entry program companion 缺少字符串字段: {field}")
    return text


def _require_bool(value: object, field: str) -> bool:
    if not isinstance(value, bool):
        raise ValueError(f"product entry program companion 缺少布尔字段: {field}")
    return value


def _require_mapping(value: object, field: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise ValueError(f"product entry program companion 缺少对象字段: {field}")
    return value


def _require_string_list(value: object, field: str) -> list[str]:
    if not isinstance(value, list):
        raise ValueError(f"product entry program companion 缺少数组字段: {field}")
    return [_require_string(item, f"{field}[{index}]") for index, item in enumerate(value)]


def _normalize_program_check(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    return {
        "check_id": _require_string(payload.get("check_id"), f"{field}.check_id"),
        "title": _require_string(payload.get("title"), f"{field}.title"),
        "status": _require_string(payload.get("status"), f"{field}.status"),
        "blocking": _require_bool(payload.get("blocking"), f"{field}.blocking"),
        "summary": _require_string(payload.get("summary"), f"{field}.summary"),
        "command": _require_string(payload.get("command"), f"{field}.command"),
    }


def _normalize_program_step(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    normalized = {
        "step_id": _require_string(payload.get("step_id"), f"{field}.step_id"),
        "command": _require_string(payload.get("command"), f"{field}.command"),
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
    }
    title = _non_empty_text(payload.get("title"))
    if title is not None:
        normalized["title"] = title
    return normalized


def _normalize_program_surface(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    command = _non_empty_text(payload.get("command"))
    ref = _non_empty_text(payload.get("ref"))
    if command is None and ref is None:
        raise ValueError("product entry program surface 至少要提供 command 或 ref。")
    normalized = {
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
    }
    if command is not None:
        normalized["command"] = command
    if ref is not None:
        normalized["ref"] = ref
    return normalized


def _normalize_guardrail_class(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    return {
        "guardrail_id": _require_string(payload.get("guardrail_id"), f"{field}.guardrail_id"),
        "trigger": _require_string(payload.get("trigger"), f"{field}.trigger"),
        "symptom": _require_string(payload.get("symptom"), f"{field}.symptom"),
        "recommended_command": _require_string(
            payload.get("recommended_command"),
            f"{field}.recommended_command",
        ),
    }


def _normalize_clearance_target(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    return {
        "target_id": _require_string(payload.get("target_id"), f"{field}.target_id"),
        "title": _require_string(payload.get("title"), f"{field}.title"),
        "commands": _require_string_list(payload.get("commands"), f"{field}.commands"),
    }


def _normalize_capability(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    return {
        "capability_id": _require_string(payload.get("capability_id"), f"{field}.capability_id"),
        "owner": _require_string(payload.get("owner"), f"{field}.owner"),
        "summary": _require_string(payload.get("summary"), f"{field}.summary"),
    }


def _normalize_workflow_coverage_item(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    return {
        "step_id": _require_string(payload.get("step_id"), f"{field}.step_id"),
        "manual_flow_label": _require_string(
            payload.get("manual_flow_label"),
            f"{field}.manual_flow_label",
        ),
        "coverage_status": _require_string(payload.get("coverage_status"), f"{field}.coverage_status"),
        "current_surface": _require_string(payload.get("current_surface"), f"{field}.current_surface"),
        "remaining_gap": _require_string(payload.get("remaining_gap"), f"{field}.remaining_gap"),
    }


def _normalize_sequence_step(value: object, field: str) -> dict[str, str]:
    payload = _require_mapping(value, field)
    normalized = {
        "step_id": _require_string(payload.get("step_id"), f"{field}.step_id"),
        "phase_id": _require_string(payload.get("phase_id"), f"{field}.phase_id"),
        "status": _require_string(payload.get("status"), f"{field}.status"),
        "summary": _require_string(payload.get("summary"), f"{field}.summary"),
    }
    title = _non_empty_text(payload.get("title"))
    if title is not None:
        normalized["title"] = title
    return normalized


def build_program_check(
    *,
    check_id: str,
    title: str,
    status: str,
    blocking: bool,
    summary: str,
    command: str,
) -> dict[str, Any]:
    return _normalize_program_check(
        {
            "check_id": check_id,
            "title": title,
            "status": status,
            "blocking": blocking,
            "summary": summary,
            "command": command,
        },
        "program_check",
    )


def build_product_entry_preflight(
    *,
    summary: str,
    recommended_check_command: str,
    recommended_start_command: str,
    checks: Sequence[Mapping[str, Any]],
) -> dict[str, Any]:
    normalized_checks = [
        _normalize_program_check(check, f"checks[{index}]")
        for index, check in enumerate(checks)
    ]
    blocking_check_ids = [
        check["check_id"]
        for check in normalized_checks
        if check["blocking"] and check["status"] != "pass"
    ]
    return {
        "surface_kind": "product_entry_preflight",
        "summary": _require_string(summary, "summary"),
        "ready_to_try_now": not blocking_check_ids,
        "recommended_check_command": _require_string(
            recommended_check_command,
            "recommended_check_command",
        ),
        "recommended_start_command": _require_string(
            recommended_start_command,
            "recommended_start_command",
        ),
        "blocking_check_ids": blocking_check_ids,
        "checks": normalized_checks,
    }


def build_workflow_coverage_item(
    *,
    step_id: str,
    manual_flow_label: str,
    coverage_status: str,
    current_surface: str,
    remaining_gap: str,
) -> dict[str, str]:
    return _normalize_workflow_coverage_item(
        {
            "step_id": step_id,
            "manual_flow_label": manual_flow_label,
            "coverage_status": coverage_status,
            "current_surface": current_surface,
            "remaining_gap": remaining_gap,
        },
        "workflow_coverage_item",
    )


def build_detailed_readiness(
    *,
    surface_kind: str,
    verdict: str,
    usable_now: bool,
    good_to_use_now: bool,
    fully_automatic: bool,
    user_experience_level: str,
    summary: str,
    recommended_start_surface: str,
    recommended_start_command: str,
    recommended_loop_surface: str,
    recommended_loop_command: str,
    workflow_coverage: Sequence[Mapping[str, Any]],
    blocking_gaps: list[str],
) -> dict[str, Any]:
    return {
        "surface_kind": _require_string(surface_kind, "surface_kind"),
        "verdict": _require_string(verdict, "verdict"),
        "usable_now": _require_bool(usable_now, "usable_now"),
        "good_to_use_now": _require_bool(good_to_use_now, "good_to_use_now"),
        "fully_automatic": _require_bool(fully_automatic, "fully_automatic"),
        "user_experience_level": _require_string(
            user_experience_level,
            "user_experience_level",
        ),
        "summary": _require_string(summary, "summary"),
        "recommended_start_surface": _require_string(
            recommended_start_surface,
            "recommended_start_surface",
        ),
        "recommended_start_command": _require_string(
            recommended_start_command,
            "recommended_start_command",
        ),
        "recommended_loop_surface": _require_string(
            recommended_loop_surface,
            "recommended_loop_surface",
        ),
        "recommended_loop_command": _require_string(
            recommended_loop_command,
            "recommended_loop_command",
        ),
        "workflow_coverage": [
            _normalize_workflow_coverage_item(item, f"workflow_coverage[{index}]")
            for index, item in enumerate(workflow_coverage)
        ],
        "blocking_gaps": _require_string_list(blocking_gaps, "blocking_gaps"),
    }


def build_guardrail_class(
    *,
    guardrail_id: str,
    trigger: str,
    symptom: str,
    recommended_command: str,
) -> dict[str, str]:
    return _normalize_guardrail_class(
        {
            "guardrail_id": guardrail_id,
            "trigger": trigger,
            "symptom": symptom,
            "recommended_command": recommended_command,
        },
        "guardrail_class",
    )


def build_product_entry_program_step(
    *,
    step_id: str,
    command: str,
    surface_kind: str,
    title: str | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "step_id": step_id,
        "command": command,
        "surface_kind": surface_kind,
    }
    if title is not None:
        payload["title"] = title
    return _normalize_program_step(payload, "program_step")


def build_product_entry_program_surface(
    *,
    surface_kind: str,
    command: str | None = None,
    ref: str | None = None,
) -> dict[str, str]:
    payload: dict[str, Any] = {"surface_kind": surface_kind}
    if command is not None:
        payload["command"] = command
    if ref is not None:
        payload["ref"] = ref
    return _normalize_program_surface(payload, "program_surface")


def build_product_entry_guardrails(
    *,
    summary: str,
    guardrail_classes: Sequence[Mapping[str, Any]],
    recovery_loop: Sequence[Mapping[str, Any]],
) -> dict[str, Any]:
    return {
        "surface_kind": "product_entry_guardrails",
        "summary": _require_string(summary, "summary"),
        "guardrail_classes": [
            _normalize_guardrail_class(item, f"guardrail_classes[{index}]")
            for index, item in enumerate(guardrail_classes)
        ],
        "recovery_loop": [
            _normalize_program_step(item, f"recovery_loop[{index}]")
            for index, item in enumerate(recovery_loop)
        ],
    }


def build_clearance_target(
    *,
    target_id: str,
    title: str,
    commands: list[str],
) -> dict[str, Any]:
    return _normalize_clearance_target(
        {
            "target_id": target_id,
            "title": title,
            "commands": commands,
        },
        "clearance_target",
    )


def build_clearance_lane(
    *,
    surface_kind: str,
    summary: str,
    recommended_step_id: str,
    recommended_command: str,
    clearance_targets: Sequence[Mapping[str, Any]],
    clearance_loop: Sequence[Mapping[str, Any]],
    proof_surfaces: Sequence[Mapping[str, Any]],
    recommended_phase_command: str,
) -> dict[str, Any]:
    return {
        "surface_kind": _require_string(surface_kind, "surface_kind"),
        "summary": _require_string(summary, "summary"),
        "recommended_step_id": _require_string(recommended_step_id, "recommended_step_id"),
        "recommended_command": _require_string(recommended_command, "recommended_command"),
        "clearance_targets": [
            _normalize_clearance_target(item, f"clearance_targets[{index}]")
            for index, item in enumerate(clearance_targets)
        ],
        "clearance_loop": [
            _normalize_program_step(item, f"clearance_loop[{index}]")
            for index, item in enumerate(clearance_loop)
        ],
        "proof_surfaces": [
            _normalize_program_surface(item, f"proof_surfaces[{index}]")
            for index, item in enumerate(proof_surfaces)
        ],
        "recommended_phase_command": _require_string(
            recommended_phase_command,
            "recommended_phase_command",
        ),
    }


def build_program_capability(
    *,
    capability_id: str,
    owner: str,
    summary: str,
) -> dict[str, str]:
    return _normalize_capability(
        {
            "capability_id": capability_id,
            "owner": owner,
            "summary": summary,
        },
        "program_capability",
    )


def build_backend_deconstruction_lane(
    *,
    summary: str,
    substrate_targets: Sequence[Mapping[str, Any]],
    backend_retained_now: list[str],
    current_backend_chain: list[str],
    optional_executor_proofs: Sequence[Mapping[str, Any]],
    promotion_rules: list[str],
    deconstruction_map_doc: str,
    recommended_phase_command: str,
    surface_kind: str = "phase4_backend_deconstruction_lane",
) -> dict[str, Any]:
    return {
        "surface_kind": _require_string(surface_kind, "surface_kind"),
        "summary": _require_string(summary, "summary"),
        "substrate_targets": [
            _normalize_capability(item, f"substrate_targets[{index}]")
            for index, item in enumerate(substrate_targets)
        ],
        "backend_retained_now": _require_string_list(
            backend_retained_now,
            "backend_retained_now",
        ),
        "current_backend_chain": _require_string_list(
            current_backend_chain,
            "current_backend_chain",
        ),
        "optional_executor_proofs": [
            dict(_require_mapping(item, f"optional_executor_proofs[{index}]"))
            for index, item in enumerate(optional_executor_proofs)
        ],
        "promotion_rules": _require_string_list(promotion_rules, "promotion_rules"),
        "deconstruction_map_doc": _require_string(
            deconstruction_map_doc,
            "deconstruction_map_doc",
        ),
        "recommended_phase_command": _require_string(
            recommended_phase_command,
            "recommended_phase_command",
        ),
    }


def build_program_sequence_step(
    *,
    step_id: str,
    phase_id: str,
    status: str,
    summary: str,
    title: str | None = None,
) -> dict[str, str]:
    payload = {
        "step_id": step_id,
        "phase_id": phase_id,
        "status": status,
        "summary": summary,
    }
    if title is not None:
        payload["title"] = title
    return _normalize_sequence_step(payload, "program_sequence_step")


def build_platform_target(
    *,
    summary: str,
    sequence_scope: str,
    current_step_id: str,
    current_readiness_summary: str,
    north_star_topology: Mapping[str, Any],
    target_internal_modules: list[str],
    landing_sequence: Sequence[Mapping[str, Any]],
    completed_step_ids: list[str],
    remaining_step_ids: list[str],
    promotion_gates: list[str],
    recommended_phase_command: str,
    surface_kind: str = "phase5_platform_target",
    land_now: list[str] | None = None,
    not_yet: list[str] | None = None,
) -> dict[str, Any]:
    normalized_landing_sequence = [
        _normalize_sequence_step(item, f"landing_sequence[{index}]")
        for index, item in enumerate(landing_sequence)
    ]
    known_step_ids = {item["step_id"] for item in normalized_landing_sequence}
    for field, values in (
        ("completed_step_ids", completed_step_ids),
        ("remaining_step_ids", remaining_step_ids),
    ):
        for step_id in _require_string_list(values, field):
            if step_id not in known_step_ids:
                raise ValueError(f"platform target {field} 必须引用 landing_sequence 中的 step_id。")
    return {
        "surface_kind": _require_string(surface_kind, "surface_kind"),
        "summary": _require_string(summary, "summary"),
        "sequence_scope": _require_string(sequence_scope, "sequence_scope"),
        "current_step_id": _require_string(current_step_id, "current_step_id"),
        "current_readiness_summary": _require_string(
            current_readiness_summary,
            "current_readiness_summary",
        ),
        "north_star_topology": dict(_require_mapping(north_star_topology, "north_star_topology")),
        "target_internal_modules": _require_string_list(
            target_internal_modules,
            "target_internal_modules",
        ),
        "landing_sequence": normalized_landing_sequence,
        "completed_step_ids": _require_string_list(completed_step_ids, "completed_step_ids"),
        "remaining_step_ids": _require_string_list(remaining_step_ids, "remaining_step_ids"),
        "promotion_gates": _require_string_list(promotion_gates, "promotion_gates"),
        "recommended_phase_command": _require_string(
            recommended_phase_command,
            "recommended_phase_command",
        ),
        **({"land_now": _require_string_list(land_now, "land_now")} if land_now is not None else {}),
        **({"not_yet": _require_string_list(not_yet, "not_yet")} if not_yet is not None else {}),
    }
